Attribute Analysis.ipynb

```diff
@@ -602,6 +606,336 @@
     "display(ui, out)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# (M2) Metric: Object occurrences and cooccurrences\n",
+    "<a id=\"metric2\"></a>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup\n",
+    "<a id=\"metric2_setup\"></a>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "categories = dataset.categories\n",
+    "names = dataset.labels_to_names\n",
+    "counts = pickle.load(open(\"results/{}/2.pkl\".format(folder_name), \"rb\"))\n",
+    "topn = 10\n",
+    "counts_mat = [np.zeros((len(categories), len(categories))) for i in range(num_attrs)]\n",
+    "for key in counts[0].keys():\n",
+    "    a, b = key.split('-')\n",
+    "    a, b = int(a), int(b)\n",
+    "    for count in range(num_attrs):\n",
+    "        counts_mat[count][b][a] = counts[count][key]\n",
+    "        counts_mat[count][a][b] = counts[count][key]\n",
+    "instance_counts = [np.diagonal(counts_mat[i]) for i in range(num_attrs)]\n",
+    "normalized = [np.divide(counts_mat[i], instance_counts[i]) for i in range(num_attrs)]\n",
+    "            \n",
+    "def make_dist(counts, total):\n",
+    "    samples = np.zeros(total)\n",
+    "    samples[:int(counts)] = 1\n",
+    "    return samples\n",
+    "# Looking at if the number of times an object appears with different attributes is statistically significant for some specific attr (analyze_attr)\n",
+    "p_values = []\n",
+    "for i in range(len(instance_counts[0])):\n",
+    "    for a in range(num_attrs):\n",
+    "        for b in range(a+1, num_attrs):\n",
+    "            p_values.append(stats.ttest_ind(make_dist(instance_counts[b][i], dataset.num_skin_tone_images[b]), make_dist(instance_counts[a][i], dataset.num_skin_tone_images[a]))[1])\n",
+    "            \n",
+    "norm_instance_counts = [instance_counts[i]/ dataset.num_skin_tone_images[i] for i in range(num_attrs)]\n",
+    "xaxis = [names[i] for i in categories]\n",
+    "barWidth = .4\n",
+    "\n",
+    "# Graphs the counts of each supercategory by gender\n",
+    "def supercategory_by_attribute():\n",
+    "    if dataset.group_mapping is not None:\n",
+    "        fontsize = 15\n",
+    "        supercategory_counts = [np.zeros(len(datasets.GROUPINGS_TO_NAMES)) for i in range(num_attrs)]\n",
+    "        for i in range(len(categories)):\n",
+    "            supercat = dataset.group_mapping(categories[i])\n",
+    "            for a in range(num_attrs):\n",
+    "                supercategory_counts[a][supercat] += norm_instance_counts[a][i]\n",
+    "        for a in range(num_attrs):\n",
+    "            supercategory_counts[a] = supercategory_counts[a][1:]\n",
```

Comment:
I believe this is assuming that supercategory [0] is people so we don't care, but now that supercategory[0] can be anything, I don't think we need this line (please verify)

Reply:
Hi, i found that we still need it. In particular, all the values of supercategory[0] printed as 1.

Reply:
yes, that's because there is a person in every image we are looking at. however, I imagine there will be datasets in the future where supercategory[0] is not person, and thus it will not all be printed as 1. also, everything being 1 is perfectly okay, so this line should indeed be removed